### PR TITLE
Flip -iree-llvm-link-embedded to true by default.

### PR DESCRIPTION
**Pull request by ScottTodd**
Tying up loose ends on https://github.com/google/iree/issues/4717.

The system linking path is more flexible for debugging and integration with platform-specific tooling, while the embedded linking path is more portable and is likely to have better performance.

Summary of changes:

* Remove explicit `--iree-llvm-link-embedded=true` from targets using `dylib-sync`
* Remove test suites that were testing the embedded path explicitly (now duplicates of the default test suites)
* Disambiguate `iree-llvm-system-linker-path`/`IREE_LLVMAOT_SYSTEM_LINKER_PATH` and `iree-llvm-embedded-linker-path`/`IREE_LLVMAOT_EMBEDDED_LINKER_PATH` flags/env-vars
  * Also remove docs referencing them. If the automatic discovery fails, error messages should give enough context for developers to debug on their own.
* Include source build of `lld` in test targets that now depend on it
* Smooth out linker tool discovery in `EmbeddedLinkerTool.cpp`

Note: this switches test coverage from system linking to embedded linking. We may want to add back some test coverage for the system linker on each platform to ensure that debugging path remains functional.

---

For history, issues encountered:

- [x] Bazel: `sh: 1: ld.lld: not found` ([logs](https://source.cloud.google.com/results/invocations/8f457f1e-229b-418d-842d-f26e57d1100a/targets/iree%2Fgcp_ubuntu%2Fbazel%2Flinux%2Fx86-swiftshader%2Fcore%2Fpresubmit/log))
- [x] riscv64: `ld.lld: error: unknown target machine` ([logs](https://source.cloud.google.com/results/invocations/7e12655d-15e3-436e-8c17-f81942cd75ca/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Friscv64%2Fpresubmit/log))
- [x] [tosa_ops/ceil.mlir](https://github.com/google/iree/blob/main/iree/test/e2e/tosa_ops/ceil.mlir): `UNAVAILABLE; ELF imports one or more symbols (trying 'ceilf');` ([logs](https://source.cloud.google.com/results/invocations/2aa34ae3-7141-48b4-91e6-d080d3980582/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-swiftshader%2Fpresubmit/log))
- [x] TF integration test for [`math_test.py`](https://github.com/google/iree/blob/main/integrations/tensorflow/e2e/math/math_test.py): `UNAVAILABLE; ELF imports one or more symbols (trying 'powf')` ([logs](https://source.cloud.google.com/results/invocations/a1f91217-7830-49c2-ae48-e46011a5571f/targets/iree%2Fgcp_ubuntu%2Fcmake-bazel%2Flinux%2Fx86-swiftshader%2Fpresubmit/log))
     - This is partially done. The librt library in LLVM target pulls in `powf` implementation from MUSL, but has some strange failures (see Issue #7415)
- [x] Test in Colab (may need to bundle `lld` with `iree-translate`, and/or ensure the tools can discover it with default flags) - [context here on Discord](https://discord.com/channels/689900678990135345/689957613152239638/900828509386178641). (Fixed by https://github.com/google/iree/pull/7436/: tested in [this notebook](https://colab.research.google.com/gist/ScottTodd/503cd4141e724c9d63dda8cf3fb3d6b2/iree-with-iree-llvm-link-embedded.ipynb))
- [ ] Write more documentation about linking modes?
- [ ] ~~General cleanup: replace now default true `linkEmbedded` flag with a new, default false, `linkPlatform` flag and check compatibility / flag ergonomics with `linkStatic` option~~

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
The `powf` should also be enabled now. Could you check?

File: iree/test/e2e/xla_ops/BUILD [hunk 1]
```diff
@@ -164,79 +164,6 @@ iree_check_single_backend_test_suite(
...
     ),
-    compiler_flags = [
-        "-iree-input-type=mhlo",
-        "-iree-llvm-link-embedded=true",
-    ],
-    driver = "dylib",
-    target_backend = "dylib-llvm-aot",
-)
-
-iree_check_single_backend_test_suite(
```
Comment (MaheshRavishankar) [inline iree/test/e2e/xla_ops/BUILD:175]: Do we want to remove this? I would rather have a test suite with the `iree-llvm-link-embedded=false`. AFAIK we want to have the `false` path for triaging/debugging purposes.

**Review by MaheshRavishankar — COMMENTED** [review]
Just one question.

File: iree/test/e2e/xla_ops/BUILD (hunk 1 above)
Comment (ScottTodd) [reply iree/test/e2e/xla_ops/BUILD:175]: I was wondering about that too:

> Note: this switches test coverage from system linking to embedded linking. We may want to add back some test coverage for the system linker on each platform to ensure that debugging path remains functional.

I don't think iree/test/e2e/xla_ops/ (op specific integration tests) is the right place for such coverage though.

Comment (ScottTodd) [reply iree/test/e2e/xla_ops/BUILD:175]: How's this? I added back a smoketest in the compiler unit tests. I think the system linker's coverage should be a superset of the embedded linker that uses librt, so this should still be sufficient test coverage? We won't really have coverage for peculiarities around system linker flags (`-nostdlib`, `-shared`, etc.), but that might be okay?

Comment (MaheshRavishankar) [reply iree/test/e2e/xla_ops/BUILD:175]: Ill let you make the call on this. I am fine with your approach.

**Review by MaheshRavishankar — APPROVED** [review]

**Review by benvanik — APPROVED** [review]

**Pull request merged**